### PR TITLE
Cache invalidation associations and blacklist

### PR DESCRIPTION
**Pull request by juanca**
**Summary**

Modifies existing control flow to be more restrictive.

Blacklist attributes are never sent to the server (hence, blacklisted). Because the server is not being updated by these modified values, the changes theoretically do not impact server results. The cache can still be respected on subsequent fetches. Example of blacklist attributes: associations, calculated fields.

**Test plan**

Integration tests showcase impact on caches when setting values on existing models. 

- [ ] https://github.com/mavenlink/mavenlink/pull/9305

### CODE REVIEW

**Comment by juanca** [comment]
I'll block this on changing `brainstem-js` as a peer dependency. Will make it easier to test!

File: src/storage-manager.coffee [hunk 1]
```diff
@@ -205,8 +203,19 @@ class _StorageManager
   #
   # Private
 
-  _invalidateCache: (name) ->
-    @collections[name].cache = {}
+  _shouldInvalidateCache: (model) ->
+    return false if !model
+    return false if model.isNew()
```
Comment (naganowl) [inline src/storage-manager.coffee]: Can these lines be consolidated into

`return false if model?.isNew()`

File: src/storage-manager.coffee [hunk 2]
```diff
@@ -205,8 +203,19 @@ class _StorageManager
...
+    return true if Object.keys(model.changed).length is 0 # destroyed
```
Comment (naganowl) [inline src/storage-manager.coffee]: This can check `model.hasChanged()` since it's logically equivalent.

File: src/storage-manager.coffee [hunk 3]
```diff
@@ -205,8 +203,19 @@ class _StorageManager
...
+
+    blacklist = model.defaultJSONBlacklist() || []
+    for attribute of model.changed when attribute not in blacklist
```
Comment (naganowl) [inline src/storage-manager.coffee]: Food for thought: Underscore can be leveraged to give

```
_.some model.changed, _.contains.bind(null, blacklist)
```

File: spec/helpers/models/time-entry.coffee [hunk 4]
```diff
@@ -11,5 +11,7 @@ class TimeEntry extends Model
     task: "tasks"
     user: "users"
 
+  defaultJSONBlacklist: ->
+    ['calculated_attribute']
```
Comment (naganowl) [inline spec/helpers/models/time-entry.coffee]: Is this the only model with a blacklisted association?

File: src/storage-manager.coffee (hunk 2 above)
Comment (juanca) [reply src/storage-manager.coffee]: Good call!

File: src/storage-manager.coffee (hunk 1 above)
Comment (juanca) [reply src/storage-manager.coffee]: No because it doesn't check for a falsey `model` value.

File: spec/helpers/models/time-entry.coffee (hunk 4 above)
Comment (juanca) [reply spec/helpers/models/time-entry.coffee]: This is just checking for a blacklisted attribute -- not necessarily an association.

Can you elaborate at what you're trying to point out?

Comment (naganowl) [reply spec/helpers/models/time-entry.coffee]: Just wondering if this change is motivated by a feature or was it something that's been broken. Typically, bigmaven API doesn't play well if you send blacklisted attributes over.

Comment (juanca) [reply spec/helpers/models/time-entry.coffee]: Exactly. This changes takes the assumption that blacklisted attributes are never sent over to the server.

**Comment by juanca** [comment]
It looks like cache invalidation is incompatible with our Backgrid integration. I'm reverting all these changes.

**Comment by naganowl** [comment]
Can you elaborate on what in Backgrid makes this incompatible?

**Pull request closed without merging**